### PR TITLE
[WIP] Pivot wider

### LINKED ISSUE #434: [ENH] pivot_wider() and pivot_longer()
# Brief Description

Pandas has several functions for pivoting. The `melt` function takes a 'wide' dataframe and makes it 'long', and `pivot_table` takes a 'long' dataframe and makes it 'wide'. Similarly, `stack` and `unstack` perform pivot operations using multilevel indices. Tom Augspurger explains all of the pandas pivoting functions in his [Modern Pandas guide](https://tomaugspurger.github.io/modern-5-tidy). If you look through his guide, you can see that the API for these four functions are all different and difficult to remember.

I would like to propose `pivot_wider` and `pivot_longer` functions inspired by functions in the R [tidyr package](https://tidyr.tidyverse.org/dev/articles/pivot.html). These functions are pure syntactic sugar around pandas `melt` and `unpivot`. They should have consistent APIs, and they should be symmetric. That is, you should be able to take a 'wide' dataframe, pass it through `pivot_longer`, then pass it through `pivot_wider` to get back to the exact same 'wide' dataframe that we started with.

# Example API

Imagine that we have a 'wide' table of heartrate data for patients treated under two different drugs, *a* and *b*.

| name | a | b |
| --- | --- | --- |
| Wilbur | 67 | 56 |
| Petunia | 80 | 90 |
| Gregory | 64 | 50 |

Here's how we would convert this to a 'long' table.

```python
df.pivot_longer(column_names=['a', 'b'], names_to=['drug'], values_to=['heartrate'])
```

The output would be a 'long' table that look like this.

| name | drug | heartrate |
|---|---|---|
| Wilbur | a   |     67  |
| Petunia |   a  |      80 |
| Gregory |   a  |      64 |
| Wilbur  |   b   |    56 |
| Petunia |    b  |    90 |
| Gregory  |  b   |     50 | 

Now we'll take the above table and transform it back into a 'wide' table using `pivot_wider`.

```python
df.pivot_wider(names_from=['drug'], values_from=['heartrate'])
```

And now we're back to our 'wide' table:

| name | a | b |
| --- | --- | --- |
| Wilbur | 67 | 56 |
| Petunia | 80 | 90 |
| Gregory | 64 | 50 |

I would love some feedback on the API before I start implementing this in code!

### PR DESCRIPTION
**Pull request by samukweku**
Please describe the changes proposed in the pull request:

- Add pivot_wider function
- Convenience arguments to collapse levels 
- Allows for easy transformation back to long form and matches original form for a number of cases
- Option to not collapse levels and retain MultiIndex form.

**This PR resolves #434 .**

# PR Checklist

Please ensure that you have done the following:

1. [x] PR in from a fork off your branch. Do not PR from `<your_username>`:`dev`, but rather from `<your_username>`:`<feature-branch_name>`.

2. [x] If you're not on the contributors list, add yourself to `AUTHORS.rst`.

3. [x] Add a line to `CHANGELOG.rst` under the latest version header (i.e. the one that is "on deck") describing the contribution.
    - Do use some discretion here; if there are multiple PRs that are related, keep them in a single line.

## Quick Check

To do a very quick check that everything is correct, follow these steps below:

- [x] Run the command `make check` from pyjanitor's top-level directory. This will automatically run:
    - black formatting
    - flake8 checking
    - running the test suite
    - docs build

Once done, please check off the check-box above.

If `make check` does not work for you, you can execute the commands listed in the Makefile individually.

## Code Changes

If you are adding code changes, please ensure the following:

- [x] Ensure that you have added tests.
- [x] Run all tests (`$ pytest .`) locally on your machine.
    - [ ] Check to ensure that test coverage covers the lines of code that you have added.
    - [ ] Ensure that all tests pass.

## Documentation Changes

If you are adding documentation changes, please ensure the following:

- [ ] Build the docs locally.
- [ ] View the docs to check that it renders correctly.

# Relevant Reviewers

Please tag maintainers to review.

- @ericmjl

### CODE REVIEW

**Comment by samukweku** [comment]
@ericmjl I am definitely missing something here; If you could point me to the error, for DeepSource; why is it having an issue with the tests?

**Comment by samukweku** [comment]
@ericmjl ; not to worry; fixed it.

File: janitor/functions.py [hunk 1]
```diff
@@ -4982,3 +4983,194 @@ def pivot_longer(
...
+        6     Bob      wk3     17
+        7     Bob      wk4     33
+        8   Carla      wk1      6
+        9   Carla      wk2     13
+        10  Carla      wk3     39
+        11  Carla      wk4     40
+
+        df = pd.DataFrame(...).pivot_wider(index="name",
+                                           names_from="variable",
+                                           values_from="value")
```
Comment (ericmjl) [inline janitor/functions.py]: Let's follow `black` style here:

```suggestion
        df = (
            pd.DataFrame(...)
            .pivot_wider(
                index="name",
                names_from="variable",
                values_from="value"
            )
```

File: janitor/functions.py [hunk 2]
```diff
@@ -4982,3 +4983,194 @@ def pivot_longer(
...
+
+            name    n  pct
+        0     1  10.0  0.1
+        1     2  20.0  0.2
+        2     3  30.0  0.3
+
+        df = (pd.DataFrame(...).assign(num=0)
+                               .pivot_wider(index='num',
+                                            names_from="name",
+                                            values_from=["n", "pct"]))
```
Comment (ericmjl) [inline janitor/functions.py]: Same `black` code style comment:

```suggestion
        df = (
            pd.DataFrame(...)
            .assign(num=0)
            .pivot_wider(
                index='num',
                names_from="name",
                values_from=["n", "pct"]
             )
         )
```

**Review by ericmjl — APPROVED** [review]
Superb work, @samukweku! I am approving. Just one small suggested change on the docs. It can be merged with or without the suggested change. 

Thanks again!

**Comment by samukweku** [comment]
Some conflicts on my end. Will have to recreate this PR

**Pull request closed without merging**